enhancements/local-storage/automatic-detection.md

```diff
@@ -0,0 +1,460 @@
+	FSType string `json:"fsType,omitempty"`
+	// If specified, a list of tolerations to pass to the discovery daemons.
+	// +optional
+	Tolerations []corev1.Toleration `json:"tolerations,omitempty"`
+	// DeviceInclusionSpec is the filtration rule for including a device in the device discovery
+	// +optional
+	DeviceInclusionSpec *DeviceInclusionSpec `json:"deviceInclusionSpec"`
+}
+
+type AutoDetectVolumeStatus struct {
+	Status string `json:"status"`
+}
+
+// DeviceMechanicalProperty holds the device's mechanical spec. It can be rotational or nonRotational
+type DeviceMechanicalProperty string
+
+const (
+	// The mechanical properties of the devices
+	// Rotational refers to magnetic disks
+	Rotational DeviceMechanicalProperty = "Rotational"
+	// NonRotational refers to ssds
+	NonRotational DeviceMechanicalProperty = "NonRotational"
+)
+
+type DeviceDiscoveryPolicyType string
+
+const (
+	// The DeviceDiscoveryPolicies that will be supported by the LSO.
+	// These Discovery policies will based on lsblk's type output
+	// Disk represents a device-type of disk
+	Disk DeviceDiscoveryPolicyType = "disk"
+	// Part represents a device-type of partion
+	Part DeviceDiscoveryPolicyType = "part"
+)
+
+type DeviceInclusionSpec struct {
+	// DeviceTypes that should be used for automatic detection. This would be one of the types supported
+	// by the local-storage operator. Currently the supported types are: disk,part
+	// If the list is empty the default value will be `[disk]`.
+	DeviceTypes []DeviceDiscoveryPolicyType `json:"deviceType"`
+
+	// DeviceMechanicalProperty denotes whether Rotational or NonRotational disks should be used.
+	// by default, it selects both
+	// +optional
+	DeviceMechanicalProperty DeviceMechanicalProperty `json:"deviceMechanicalProperty"`
+
+	// MinSize is the minimum size of the device which needs to be included
+	// +optional
+	MinSize resource.Quantity `json:"minSize"`
+
+	// MaxSize is the maximum size of the device which needs to be included
+	// +optional
+	MaxSize resource.Quantity `json:"maxSize"`
+
+	// Models is a list of device models. If not empty, the device's model as outputted by lsblk needs
+	// to contain at least one of these strings.
+	// +optional
+	Models []string `json:"models"`
+
+	// Vendors is a list of device vendors. If not empty, the device's model as outputted by lsblk needs
+	// to contain at least one of these strings.
+	// +optional
+	Vendors []string `json:"vendors"`
+}
+
+type AutoDetectVolumeStatus struct {
+	Status string `json:"status"`
+}
+```
+### Workflow of LocalPV creation via AutoDetectVolume CR
+- Once the `AutoDetectVolume` is created, the controller will:
+  - configure the local-static-provisioner to make a new StorageClass based on certain directories on the selected nodes.
+  - assign diskmaker daemons to the selected nodes.
+- The diskmaker daemon will find devices that match the disovery policy and symlink them into the directory that the local-static-provisioner is watching.
```

Comment:
How do you solve races between device discovery (via `LocalVolumeDiscovery`) and diskmaker? Both of them may be inspecting the same device at the same time and discovery job may report it as free, while diskmaker will symlink it and external-provisioner will create PV for it.

Reply:
I am fine with that, as long as you don't have expectation from GUI to get accurate results.